model/model_training/models/peft_modeling.py

```diff
@@ -0,0 +1,101 @@
+from dataclasses import dataclass
+from pathlib import Path
+
+import torch
+from huggingface_hub import hf_hub_download
+from model_training.utils.utils import get_model, get_tokenizer
+from peft import LoraConfig, PeftModel, PrefixTuningConfig, get_peft_model, prepare_model_for_int8_training
+
+
+def load_peft_model(model, peft_model_path, tokenizer):
+    model.resize_token_embeddings(len(tokenizer))
+    model.config.eos_token_id = tokenizer.eos_token_id
+    model.config.bos_token_id = tokenizer.bos_token_id
+    model.config.pad_token_id = tokenizer.pad_token_id
+    model = PeftModel.from_pretrained(
+        model,
+        peft_model_path,
+        torch_dtype=model.dtype,
+    )
+    model.eos_token_id = tokenizer.eos_token_id
+    extra_embeds = hf_hub_download(peft_model_path, "extra_embeddings.pt")
+    embed_weights = torch.load(extra_embeds, map_location=model.device)
+    model.base_model.model.model.embed_tokens.weight[len(tokenizer) - embed_weights.shape[0] :, :] = embed_weights.to(
+        model.base_model.model.model.embed_tokens.weight.dtype
+    )
+    return model
+
+
+def prepare_model_for_gradient_checkpointing(model):
+    r"""
+    Prepares the model for gradient checkpointing if necessary
+    """
+    if not getattr(model, "is_loaded_in_8bit", False):
+        if hasattr(model, "enable_input_require_grads"):
+            model.enable_input_require_grads()
+        else:
+
+            def make_inputs_require_grad(module, input, output):
+                output.requires_grad_(True)
+
+            model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)
+    return model
+
+
+def peft_model(model, peft_type="lora", int8_training=False, gradient_checkpointing=False):
+    if peft_type == "lora":
+        config = LoraConfig(
+            r=16,
+            lora_alpha=32,
+            target_modules=["q_proj", "k_proj", "v_proj", "o_proj"],
+            lora_dropout=0.05,
+            bias="none",
+            task_type="CAUSAL_LM",
+        )
+    elif peft_type == "prefix-tuning":
+        config = PrefixTuningConfig(
+            num_virtual_tokens=30, prefix_projection=True, encoder_hidden_size=1024, task_type="CAUSAL_LM"
+        )
+    else:
+        raise ValueError("peft_method config is lora or prefix-tuning")
+    model = get_peft_model(model, config)
+    if int8_training:
+        model = prepare_model_for_int8_training(model)
+
+    if gradient_checkpointing:
+        model = prepare_model_for_gradient_checkpointing(model)
+    model.print_trainable_parameters()
+    return model
+
+
+@dataclass
+class SaveLoraConfig:
+    dtype: torch.dtype = torch.float16
+    is_reward_model: bool = False
+    quantization: bool = False
+    seq2seqmodel: bool = False
+    freeze_layer: bool = False
+    residual_dropout: float = 0
+    use_flash_attention: bool = False
+    adapter_save_path: str = "adapter_13B_new"
```

Comment:
Does this only work on 13B model? Should it be hardcoded?

Reply:
it's the output path for a directory with the adapter weights in them. I've changed to just "adapter"

Reply:
@smeyerhot I've updated the configs.